hearbaseline/vqt.py

```diff
@@ -0,0 +1,140 @@
+"""
+variable-Q transform spectrogram (60 bins/octave, ~10ms step)
+
+Based upon DCASE 2016 Task 2 baseline.
+The scene embeddings will be an average, i.e. only 84 dimensions.
+"""
+
+import math
+import numpy as np
+from typing import Tuple
+
+import librosa
+import torch
+from torch import Tensor
+
+
+class VQT(torch.nn.Module):
+    # Not really a torch module
+
+    # sample rate and embedding sizes are required model attributes for the HEAR API
+    sample_rate = 44100
+    # dcase baseline was 10 ms (441), but for librosa hop_length must be
+    # a positive integer multiple of 2^8 for 9-octave CQT/VQT
+    hop_length = 512
+    gamma = 30
+    bins_per_octave = 60
+    fmin = 27.5
+    noctaves = int(math.log2(sample_rate / 2 / fmin))
+
+    embedding_size = bins_per_octave * noctaves
+    scene_embedding_size = embedding_size
+    timestamp_embedding_size = embedding_size
+
+    def __init__(self):
+        super().__init__()
+
+    def forward(self, x: Tensor):
+        embeddings = []
+        for i in range(x.shape[0]):
+            embedding = np.abs(
+                librosa.vqt(
+                    x[i].numpy(),
+                    sr=self.sample_rate,
+                    hop_length=self.hop_length,
+                    gamma=self.gamma,
+                    n_bins=self.embedding_size,
+                    bins_per_octave=self.bins_per_octave,
+                    fmin=self.fmin,
+                )
+            ).T
+            embedding = torch.tensor(embedding, device=x.device)
+            embeddings.append(embedding)
+        embeddings = torch.stack(embeddings)
+        return embeddings
```

Comment:
An additional step in the DCASE baseline is some filtering (remove some low freq bins and emphasizing some higher ones) and removing background noise, if we want to be more dcase-y:

```
% Emphasize high freqs and remove low frequency bins
f = X(:,21:end)';
f = f .* repmat( linspace( 1, 15, size( f, 1))', 1, size( f, 2));
X =f';


% Remove background noise from input
p1 = prctile(sum(X'),5);
p2 = prctile(sum(X'),10);
silentFrames = find(sum(X')>p1 & sum(X')<p2);
silentTemplate = mean(X(silentFrames,:));
noiseLevel = repmat(silentTemplate,[size(X,1) 1]);
Y = max(X - noiseLevel,0);
fprintf('%s','done');
fprintf('\n');
```

Reply:
Good to know. Let's consider it a little later